evm/src/cpu/kernel/tests/bignum.rs

```diff
@@ -0,0 +1,376 @@
+
+    // Compare.
+    assert_eq!(actual_result, expected_result);
+
+    Ok(())
+}
+
+fn test_mul_bignum<F>(prepare_two_bignums_fn: &F) -> Result<()>
+where
+    F: Fn(usize) -> (BigUint, BigUint, U256, Vec<U256>),
+{
+    let (a, b, length, memory) = prepare_two_bignums_fn(1000);
+
+    // Determine expected product.
+    let product = a * b;
+    let expected_product: Vec<U256> = biguint_to_mem_vec(product);
+
+    // Output and scratch space locations (initialized as zeroes) follow a and b in memory.
+    let a_start_loc = 0.into();
+    let b_start_loc = length;
+    let output_loc = length * 2;
+
+    // Prepare stack.
+    let retdest = 0xDEADBEEFu32.into();
+    let mut initial_stack: Vec<U256> = vec![length, a_start_loc, b_start_loc, output_loc, retdest];
+    initial_stack.reverse();
+
+    // Prepare interpreter.
+    let mul_bignum = KERNEL.global_labels["mul_bignum"];
+    let mut interpreter = Interpreter::new_with_kernel(mul_bignum, initial_stack);
+    interpreter.set_kernel_general_memory(memory);
+
+    // Run mul function.
+    interpreter.run()?;
+
+    // Determine actual product.
+    let new_memory = interpreter.get_kernel_general_memory();
+    let output_location: usize = output_loc.try_into().unwrap();
+    let actual_product: Vec<_> =
+        new_memory[output_location..output_location + expected_product.len()].into();
+
+    assert_eq!(actual_product, expected_product);
+
+    Ok(())
+}
+
+#[test]
+fn test_shr_bignum_all() -> Result<()> {
+    test_shr_bignum(&prepare_bignum_random)?;
+    test_shr_bignum(&prepare_bignum_max)?;
+    test_shr_bignum(&prepare_bignum_min)?;
+
+    Ok(())
+}
+
+#[test]
+fn test_iszero_bignum_all() -> Result<()> {
+    test_iszero_bignum(&prepare_bignum_random)?;
+    test_iszero_bignum(&prepare_bignum_max)?;
+    // No need to test for min, since it is zero.
+
+    Ok(())
+}
+
+#[test]
+fn test_ge_bignum_all() -> Result<()> {
+    test_ge_bignum(&prepare_two_bignums_random)?;
+    test_ge_bignum(&prepare_two_bignums_max)?;
+    test_ge_bignum(&prepare_two_bignums_min)?;
+    test_ge_bignum(&prepare_two_bignums_diff)?;
+
+    Ok(())
+}
+
+#[test]
+fn test_add_bignum_all() -> Result<()> {
+    test_add_bignum(&prepare_two_bignums_random)?;
+    test_add_bignum(&prepare_two_bignums_max)?;
+    test_add_bignum(&prepare_two_bignums_min)?;
+    test_add_bignum(&prepare_two_bignums_diff)?;
+
+    Ok(())
+}
+
+#[test]
+fn test_addmul_bignum_all() -> Result<()> {
+    test_addmul_bignum(&prepare_two_bignums_random)?;
+    test_addmul_bignum(&prepare_two_bignums_max)?;
+    test_addmul_bignum(&prepare_two_bignums_min)?;
+    test_addmul_bignum(&prepare_two_bignums_diff)?;
+
+    Ok(())
+}
+
+#[test]
+fn test_mul_bignum_all() -> Result<()> {
+    test_mul_bignum(&prepare_two_bignums_random)?;
+    test_mul_bignum(&prepare_two_bignums_max)?;
+    test_mul_bignum(&prepare_two_bignums_min)?;
+    test_mul_bignum(&prepare_two_bignums_diff)?;
+
+    Ok(())
```

Comment:
I added a function
```
fn prepare_two_bignums_zero(_bit_size: usize) -> (BigUint, BigUint, U256, Vec<U256>) {
    let a = BigUint::zero();
    let b = BigUint::zero();
    let length: U256 = bignum_len(&a).into();
    let memory = pack_bignums(&[a.clone(), b.clone()], length.try_into().unwrap());

    (a, b, length, memory)
}
```
and called it as `test_{add,mul,addmul}_bignum(&prepare_two_bignums_zero)` in the tests above. This causes them to hang (assuming you've made my suggested change to `bignum_len` above), presumably because they don't handle `len == 0`.